File: Phan_B/baitap43/baitap43.py
import sys

dem = 0

def fib_de_quy(n: int) -> int:
    global dem
    dem += 1
    if n <= 2:
        return 1
    return fib_de_quy(n - 1) + fib_de_quy(n - 2)

def fib_ghi_nho(n: int, bo_nho=None) -> int:
    if bo_nho is None:
        bo_nho = {}
    if n <= 2:
        return 1
    if n in bo_nho:
        return bo_nho[n]
    bo_nho[n] = fib_ghi_nho(n - 1, bo_nho) + fib_ghi_nho(n - 2, bo_nho)
    return bo_nho[n]

def fib_lap(n: int) -> int:
    if n <= 2:
        return 1
    a, b = 1, 1
    for _ in range(3, n + 1):
        a, b = b, a + b
    return b

def main():
    global dem
    content = sys.stdin.read().split()
    if not content:
        return
    for token in content:
        n = int(token)
        if n >= 93:
            print("tran long long - phai phat hien va bao", flush=True)
            continue

        if n <= 35:
            dem = 0
            ans = fib_de_quy(n)
            print(f"F = {ans}, so loi goi = {dem}", flush=True)
        else:
            ans = fib_lap(n)
            print(f"F = {ans}", flush=True)

if __name__ == "__main__":
    main()

File: Phan_B/baitap43/baitap43.jl
const DEM = Ref(0)

function fib_de_quy(n)
    DEM[] += 1
    if n <= 2
        return Int64(1)
    end
    return fib_de_quy(n - 1) + fib_de_quy(n - 2)
end

function fib_ghi_nho(n, memo=Dict{Int, Int64}())
    if n <= 2
        return Int64(1)
    end
    if haskey(memo, n)
        return memo[n]
    end
    memo[n] = fib_ghi_nho(n - 1, memo) + fib_ghi_nho(n - 2, memo)
    return memo[n]
end

function fib_lap(n)
    if n <= 2
        return Int64(1)
    end
    a = Int64(1)
    b = Int64(1)
    for i in 3:n
        c = a + b
        a = b
        b = c
    end
    return b
end

function main()
    content = read(stdin, String)
    tokens = split(content)
    if isempty(tokens) return end

    for token in tokens
        n = parse(Int, token)
        if n >= 93
            println("tran long long - phai phat hien va bao")
            continue
        end

        if n <= 35
            DEM[] = 0
            ans = fib_de_quy(n)
            println("F = $(ans), so loi goi = $(DEM[])")
        else
            ans = fib_lap(n)
            println("F = $(ans)")
        end
        flush(stdout)
    end
end

main()